File: hashtable/202/202.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
  let set = new Set();
  while (n !== 1 && !set.has(n)) {
    set.add(n);
    n = getNext(n);
  }
  return n === 1;
};

const getNext = n => {
  let sum = 0;
  while (n > 0) {
    sum += (n % 10) * (n % 10);
    n = Math.floor(n / 10);
  }
  return sum;
};

File: hashtable/202/202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def getNext(n):
            sum = 0
            while n > 0:
                n, remainder = divmod(n, 10)
                sum += remainder * remainder
            return sum

        hashset = set()
        while n != 1 and n not in hashset:
            hashset.add(n)
            n = getNext(n)
        return n == 1

